signac/contrib/project.py

```diff
@@ -1248,15 +1278,21 @@ def _get_statepoint(self, job_id, fn=None):
                     )
                     self._sp_cache_warned = True
                 statepoint = self._get_statepoint_from_workspace(job_id)
+                # Update the project's state point cache from this cache miss
+                self._sp_cache[job_id] = statepoint
         except KeyError as error:
+            # Fall back to a file containing all state points because the state
+            # point could not be read from the job workspace.
             try:
-                statepoint = self.read_statepoints(fn=fn)[job_id]
+                statepoints = self.read_statepoints(fn=fn)
+                # Update the project's state point cache
+                self._sp_cache.update(statepoints)
+                statepoint = statepoints[job_id]
```

Comment:
A more aggressive solution would be to update the internal state point cache inside the methods `_get_statepoint_from_workspace` and `read_statepoints`. I tested this and all of our existing tests pass -- I think this would be valid. In every case where we currently call those methods, we manually update the cache with the results. Moving that "read to cache" logic into the implementation of those methods makes sense to me, but I would like a second opinion. That can also be done in a follow-up PR.

Reply:
Since this PR seems almost done, my hunch would be to leave the aggressive solution for a follow-up PR. But I'm still understanding this PR and can comment better when I finish my review.